### PR TITLE
Material dialogs; FAB improvements; multiple fixes

### PR DESCRIPTION
**Pull request by jpmcosta**
Main feature:
- backport of Material dialogs for pre-L androids.

Side improvements:
- Adds `RippleMaterialDrawable` – successor to `RippleDrawableSimpleCompat`. Still a very simple drawable, but a good replacement to Lollipop's `RippleDrawable`. Can, and will, be improved in the future.;
- `LayerMaterialDrawable` improvements – adds constant state. A needed improvement to support `RippleMaterialDrawable`;
- `FloatingActionButton` improvements;
- Adds `GradientMaterialDrawable`;
- Updates buttons' drawables;
- Other fixes.

Known issues:
- backport of Material dialogs lacks background shadow.

### CODE REVIEW

File: MaterialLibrary/src/main/res/layout/alert_dialog_material.xml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="vertical">
+        <LinearLayout android:id="@+id/title_template"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="horizontal"
+            android:gravity="center_vertical|start"
+            android:paddingStart="@dimen/alert_dialog_padding_material"
+            android:paddingEnd="@dimen/alert_dialog_padding_material"
```
Comment (rastislavvasko) [inline MaterialLibrary/src/main/res/layout/alert_dialog_material.xml]: This should be `android:paddingLeft` and `android:paddingRight`.

Comment (jpmcosta) [inline MaterialLibrary/src/main/res/layout/alert_dialog_material.xml]: Nice catch. Thank you!

**Comment by goncalossilva** [comment]
Another issue: when the "Google Play Services is missing" dialog kicks in in API 16, I get this exception:

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.todoist.debug/com.todoist.activity.HomeActivity}: android.content.res.Resources$NotFoundException: File res/drawable/dialog_background_material.xml from drawable resource ID #0x7f0200a0
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2059)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: android.content.res.Resources$NotFoundException: File res/drawable/dialog_background_material.xml from drawable resource ID #0x7f0200a0
            at android.content.res.Resources.loadDrawable(Resources.java:1918)
            at android.content.res.Resources.getDrawable(Resources.java:659)
            at com.android.internal.policy.impl.PhoneWindow.generateLayout(PhoneWindow.java:2804)
            at com.android.internal.policy.impl.PhoneWindow.installDecor(PhoneWindow.java:2844)
            at com.android.internal.policy.impl.PhoneWindow.setContentView(PhoneWindow.java:252)
            at com.android.internal.app.AlertController.installContent(AlertController.java:240)
            at android.app.AlertDialog.onCreate(AlertDialog.java:336)
            at android.app.Dialog.dispatchOnCreate(Dialog.java:351)
            at android.app.Dialog.show(Dialog.java:256)
            at android.support.v4.app.DialogFragment.onStart(DialogFragment.java:398)
            at android.support.v4.app.Fragment.performStart(Fragment.java:1810)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:987)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1136)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1118)
            at android.support.v4.app.FragmentManagerImpl.dispatchStart(FragmentManager.java:1932)
            at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:565)
            at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1163)
            at android.app.Activity.performStart(Activity.java:5018)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2032)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.UnsupportedOperationException: Can't convert to color: type=0x2
            at android.content.res.TypedArray.getColor(TypedArray.java:326)
            at android.graphics.drawable.GradientDrawable.inflate(GradientDrawable.java:951)
            at android.graphics.drawable.Drawable.createFromXmlInner(Drawable.java:881)
            at android.graphics.drawable.InsetDrawable.inflate(InsetDrawable.java:113)
            at android.graphics.drawable.Drawable.createFromXmlInner(Drawable.java:881)
            at android.graphics.drawable.Drawable.createFromXml(Drawable.java:818)
            at android.content.res.Resources.loadDrawable(Resources.java:1915)
            at android.content.res.Resources.getDrawable(Resources.java:659)
            at com.android.internal.policy.impl.PhoneWindow.generateLayout(PhoneWindow.java:2804)
            at com.android.internal.policy.impl.PhoneWindow.installDecor(PhoneWindow.java:2844)
            at com.android.internal.policy.impl.PhoneWindow.setContentView(PhoneWindow.java:252)
            at com.android.internal.app.AlertController.installContent(AlertController.java:240)
            at android.app.AlertDialog.onCreate(AlertDialog.java:336)
            at android.app.Dialog.dispatchOnCreate(Dialog.java:351)
            at android.app.Dialog.show(Dialog.java:256)
            at android.support.v4.app.DialogFragment.onStart(DialogFragment.java:398)
            at android.support.v4.app.Fragment.performStart(Fragment.java:1810)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:987)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1136)
            at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1118)
            at android.support.v4.app.FragmentManagerImpl.dispatchStart(FragmentManager.java:1932)
            at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:565)
            at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1163)
            at android.app.Activity.performStart(Activity.java:5018)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2032)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
```

Looking at `dialog_background_material.xml`, the problem is the `<solid android:color`. If I change it to some specific color it works.

**Comment by goncalossilva** [comment]
Hum... it's not the only problem. I will look more into this.

**Comment by goncalossilva** [comment]
False alarm :smile: My branch was dirty and I was using an older version!

**Pull request merged**